Adds arbitrary date parsing

tkroman opened the pull request:
Inspired by a recent discussion @ slack channel, where a user was complaining that he can't compare dates stored as strings in a way other than lexicographical.
This PR adds a procedure that accepts a (string) value and an optional `DateFormat`-compatible pattern and returns a unixtime of date parsed from value with the given format. null input => null output.

Real-life example:

```
MATCH (p:Person) 
WITH p 
CALL apoc.date.unixtime(p.born) 
YIELD value AS dob
RETURN p 
ORDER BY dob
```

will order people by dates given that their 'born' field are compatible with the default `EEE MMM dd HH:mm:ss zzz yyyy` format.

Since procedure methods can't be overloaded, there's a difference in names: `unixtime(date)` vs `unixtime_f(date, format)`, which is totally disputable.

jexp commented on src/main/java/apoc/date/Date.java:
```diff
@@ -0,0 +1,43 @@
+
+	private static DateFormat getFormat(String pattern) {
+		if (pattern == null) {
+			return new SimpleDateFormat("EEE MMM dd HH:mm:ss zzz yyyy");
```
Should we use the ISO8601 time format as default? or sth like `"yyyy-mm-dd HH:MM:SS"` ?

jexp commented on src/test/java/apoc/date/DateTest.java:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	@Test public void testDateAccessor() throws Exception {
+		try (Transaction tx = db.beginTx()) {
```
you can simplify the test by just passing in the string value a parameter value. No need to create nodes and such.

jexp commented:
It would be cool to have the inverse function (unix-time->long) too.

And perhaps something that splits either date type into a map format with entries for all parts.

Could you add the section / function(s) to the readme too?

tkroman commented:
@jexp, I'll simplify tests and add docs.
On account of default format, "yyyy-mm-dd HH:MM:SS" seems reasonable, although I choose the current one on account that it's the one produced by `Date#toString()`, so I'm not 100% sure.
On account of the 

> inverse function (unix-time->long) too.

do you mean unixtime to date?

jexp commented:
we shouldn't care about Java internal formattings :) neither in Cypher nor Neo 

yes unixdate to string date

tkroman commented:
@jexp I addressed your comments, all except the one about parts. I'm not sure how to approach it in a universal way (looks like there's no nice way to iterate through `Calendar`'s values), so I guess, I'll have to think on that a little.

jexp commented:
I'd like to merge it, but love to have nicer names.
(Also the names for procedures are camel-case not lower-case-underscore).

I love if the names flow, like `apoc.date.parse` or `apoc.date.toString`, `apoc.date.from/toUnix`,`apoc.date.from/toMillis, or `apoc.date.unixToString`.

I'm also not sure if we should put "unix" in the name? Perhaps rather seconds/millis (and document that they are since Jan 1st 1970). Btw. what for parse / format we need to set a locale and timezone too. I'm not sure that it defaults to UTC, I think it defaults to the local TZ.

tkroman commented:
@jexp, I've addressed all the comments.
The only conceptually questionable thing is the `apoc.date.listFields` now. I decided not to anchor pattern names to values (i.e. `yyyy` => `2016`, `MM` => `Jan`), but rather use a universal `Months`, `Seconds` etc.

jexp commented:
Amazing thanks so much. Looks great. I'll change listFields to just fields and lowercase the field names.

jexp commented:
merged manually
